petastorm/pytorch.py

```diff
@@ -367,3 +367,123 @@ def __enter__(self):
     def __exit__(self, exc_type, exc_val, exc_tb):
         self.reader.stop()
         self.reader.join()
+
+
+class InMemBatchedDataLoader(LoaderBase):
+    """
+    Same as BatchedDataLoader except it only loads upto capacity rows into memory.
+    """
+
+    def __init__(self, reader, batch_size=1,
+                 transform_fn=None,
+                 num_epochs=1,
+                 seed=0,
+                 capacity=1024,
+                 shuffle=False):
+        """
+        Initializes a data loader object.
+        This class does not support special types that are not supported in PyTorch (decimal/string).
+        :param reader: petastorm Reader instance
+        :param batch_size: the number of items to return per batch; factored into the len() of this reader
+        :param transform_fn: an optional callable to convert batches from the reader to PyTorch tensors
+        :param num_epochs: number of epochs.
+        :param seed: random seed used to shuffle.
+        :param capacity: number of rows to be loaded into memory.
+        :param shuffle: If ``True``, indices will be shuffled in every epoch.
+        """
+        super(InMemBatchedDataLoader, self).__init__()
+        self.reader = reader
+        self.batch_size = batch_size
+        self.transform_fn = transform_fn or torch.as_tensor
+        self.num_epochs = num_epochs
+        self.seed = seed
+        self.capacity = capacity
+        self.shuffle = shuffle
+        self.size = 0
+
+    def _load_rows_into_mem(self):
+        n_rows = 0
+        buffer_full = False
+        self.buffer = None
+        self.keys = None
+
+        for row in self.reader:
+            if buffer_full:
+                break
```

Comment:
It's easy for a user to assume that if they use this class all data going to be loaded into memory. They might not notice the `capacity` constructor argument since it has a default value. I think we should make `capacity` a mandatory argument or find some other API change that make this behavior of dropping rows more obvious. 

Perhaps tweaking the name to hint that the value is given as rows (and not bytes), e.g. rows_capacity.